src/rsa.py

```diff
@@ -80,45 +80,60 @@ def sign(body, user):
+                                             mode='wb') as tmpfile:
+                tmpfile_path = tmpfile.name
+
             cmd = [ "openssl", "dgst", "-sign",
                    os.path.join(RSA_DIR, key_id, "private.pem"),
                    "-keyform", "PEM",
-                   "-sha256", "-out", tmpfile, tmpinput ]
-            cmd_result = subprocess.run(cmd, check=False)
-            if cmd_result.returncode != 0:
-                raise RuntimeError("Failed to sign payload")
-            tmpout = tempfile.mkstemp(prefix="rsa-")[1]
-            with open(tmpout, 'a', encoding='utf-8') as output:
-                output.write(sha256sum(tmpinput))
-                output.write("\n")
-                output.write("ts: " +
+                   "-sha256", "-out", tmpfile_path, tmpinput_path ]
+            subprocess.run(cmd, check=True)
+
+            with tempfile.NamedTemporaryFile(prefix="rsa-", delete=False,
+                                             mode='w', encoding='utf-8') as tmpout:
+                tmpout_path = tmpout.name
+                tmpout.write(sha256sum(tmpinput_path))
+                tmpout.write("\n")
+                tmpout.write("ts: " +
                              str(int(datetime.utcnow().timestamp())))
-                output.write("\n")
+                tmpout.write("\n")
                 hex_string=""
-                with open(tmpfile, 'rb') as ifile:
+                with open(tmpfile_path, 'rb') as ifile:
                     while True:
                         chunk = ifile.read(4096)
                         if not chunk:
                             break
                         hex_string += binascii.hexlify(chunk).decode('utf-8')
-                output.write("rsa2048: " + hex_string)
-                output.write("\n")
+                tmpout.write("rsa2048: " + hex_string)
+                tmpout.write("\n")
 
-            with open(tmpout, 'rb') as signature:
+            with open(tmpout_path, 'rb') as signature:
                 response = {
                     "signature": binascii.b2a_base64(signature.read()).decode().rstrip("\n")
                 }
 
                 LOG.info("%s successfully signed a payload using %s key", user, key_id)
-
-        os.remove(tmpinput)
-        os.remove(tmpout)
-        os.remove(tmpfile)
-        return response
+                return response
+        except subprocess.CalledProcessError as excp:
+            LOG.error("Subprocess failed: %s", excp)
+            return {"error": "Failed to sign payload"}, 500
+        except Exception as excp:
+            LOG.error("Unexpected error: %s", excp)
+            return {"error": str(excp)}, 500
+        finally:
+            if os.path.exists(tmpinput_path):
+                os.remove(tmpinput_path)
+            if os.path.exists(tmpinput_path):
+                os.remove(tmpinput_path)
+            if os.path.exists(tmpfile_path):
```

Comment:
This does not address what I meant. Because the `*_path` variables are only first defined as you go, `tmpfile_path` can be undefined here if the exception is thrown e.g. while writing `tmpinput`.

Also this removes `tmpinput_path` twice and does not remove `tmpout_path`.